wip: clipping support

emranemran opened the pull request.

thomshutt commented on video/clip.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+// Finds the segment in an HLS manifest that contains the timestamp (aka playhead) specified.
+func getRelevantSegment(allSegments []*m3u8.MediaSegment, relevantSegments *[]*m3u8.MediaSegment, playHeadTime float64, startingIdx uint64) (segNum uint64) {
```
Why is relevantSegments an input and not an output?

thomshutt replied:
I think you can do it all with indexes at the point, so something like `getTimestampSegmentIndex(allSegments []*m3u8.MediaSegment, playHeadTime float64) (int, error)`

mjh1 replied:
But maybe it would make more sense if `getRelevantSegment` just returned an index, then in `ClipManifest` it can form the `relevantSegments` slice when it knows both the start and end index.

emranemran replied:
Alright, updated code to use just indices instead of passing pointers around.

thomshutt commented on video/clip.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+// Function to calculate the total duration of the HLS manifest
+/*func calculateTotalDuration(manifest *m3u8.MediaPlaylist) float64 {
```
Can we get rid of this?

emranemran replied:
Done

mjh1 commented on video/clip.go:
```diff
@@ -0,0 +1,72 @@
+			continue
+		}
+		// Skip adding to slice of relevantSegments if the segment already exists
+		// (e.g. start and end times fall within same segment index)
```
if this is the only case where this is possible maybe we could just increment `startingIdx` by one and exit if we've gone past the `endTime`. Or check that the current index is equal to `startingIdx`. Then we could avoid looping through `relevantSegments`

emranemran replied:
Updated the logic to simplify all of this further.

emranemran commented:
**Updates**: 
* `ClipManifest` now returns a slice of `m3u8.MediaSegment` that contains **all** segments from the specified start time to end time
* `getRelevantSegment` now returns just indices and doesn't use any pointers (even though my brain sometimes prefers passing around pointers). 
* Added back `getTotalDurationAndSegments` function that's useful to simplify the range loop
* Updated tests to match the new logic

mjh1 commented on video/clip.go:
```diff
@@ -0,0 +1,90 @@
+	var totalDuration float64
+	var segmentCount uint64
+	for _, segment := range manifest.Segments {
+		if segment == nil {
+			break
+		}
```
nit: just use `manifest.GetAllSegments()` then you can avoid this if statement and just take the length of that slice for segmentCount rather than needed to calculate yourself.

emranemran replied:
Ah good point, I seem to have missed that method already exists in the m3u8 library. Will update.

emranemran replied:
Updated

mjh1 commented on video/clip.go:
```diff
@@ -0,0 +1,90 @@
+	}
+
+	// Generate a slice of all segments that overlap with startTime to endTime
+	relevantSegments := manifest.Segments[startSegIdx : endSegIdx+1]
```
it looks like this `endSegIdx+1` will be out of range if we hit the line above `endSegIdx = manifestSegments - 1` is there a unit test covering it?
